packages/frontend/src/hooks/useApi.ts

```diff
@@ -1,20 +1,140 @@
 import useSWR from 'swr';
+import { paths } from '../types/api-schema';
 
-interface UseApiArgs {
+interface UseApiArgs<
+  TUrl extends keyof paths,
+  TMethod extends keyof paths[TUrl],
+  TBodyContentType extends string = 'application/json',
+  TResponseContentType extends string = 'application/json',
+> {
   /** The method, i.e. GET, POST */
-  method?: string;
+  method?: TMethod;
   /** The body of the request */
-  body?: string;
-  /** Headers to pass onto the request, note Authorization is set by default, as is Content-Type to json. */
+  // @ts-ignore
+  body?: paths[TUrl][TMethod]['requestBody']['content'][TBodyContentType];
+  /** Headers to pass onto the request, note Authorization is set by default, as
+   * is Content-Type to json. */
   headers?: { [key: string]: string };
-  /** Query parameters to pass onto the request, to avoid having to manually add to the url. */
+  /** Query parameters to pass onto the request, to avoid having to manually add
+   * to the url. */
   queryParams?: { [key: string]: string | number };
   /** How to parse the response. Will by default parse as json. */
   parser?: (res: Response) => Promise<any>;
+  /** Path arguments, i.e. /api/{id}, id is a path parameter. They should be
+   * passed in the same order as declared in the URL. */
+  pathParams?: ObjectFromUnion<
+    AsUnion<RemoveNonVariable<SplitLiteral<TUrl, '/'>>>,
+    string | number
+  >;
+  /** The content type of the body. */
+  bodyContentType?: TBodyContentType;
+  /** The content type of the response */
+  responseContentType?: TResponseContentType;
 }
 
+/**
+ * Make a request to our API. Do not manually use the type parameters, they are
+ * used internally and should be implicitly generated based on your arguments.
+ *
+ * See {@link useUntypedApi} if the type system is causing issues.
+ *
+ * @param url the url to the api, i.e. /api/v1/ping
+ * @param options the options to pass to the request
+ * @returns the response from the api. If data is undefined, but there is no
+ * error, then it is loading.
+ *
+ * @example const {data, error} = useApi("/api/v1/ping", {method: "GET"});
+ */
+export const useApi = <
+  TUrl extends keyof paths,
+  TMethod extends keyof paths[TUrl],
+  TBodyContentType extends string = 'application/json',
+  TResponseContentType extends string = 'application/json',
+>(
+  url: TUrl,
+  {
+    method,
+    parser = (res) => res.json(),
+    pathParams,
+    ...options
+  }: UseApiArgs<TUrl, TMethod, TBodyContentType, TResponseContentType> = {},
+) => {
+  const { data, error, ...rest } = useSWR<
+    // @ts-ignore
```

Comment:
may I know why @ts-ignore is needed?

Reply:
Indexing an interface with a normal string is not generally typesafe, but in this case the generated interface is guaranteed to have those keys.